src/streams.h

```diff
@@ -802,16 +804,19 @@ class CBufferedFile
 
     //! rewind to a given reading position
     bool SetPos(uint64_t nPos) {
-        nReadPos = nPos;
-        if (nReadPos + nRewind < nSrcPos) {
-            nReadPos = nSrcPos - nRewind;
+        size_t bufsize = vchBuf.size();
+        if (nSrcPos > bufsize && nPos < nSrcPos - bufsize) {
```

Comment:
I think this is wrong, `vchBuf.size()` if constant and the actual read position is obtained from `nReadPos % vchBuf.size()` - it's a ring buffer.

Reply:
I checked this carefully, and I'm pretty sure it's right, please check again. I did simplify it slightly (without changing the logic) in the second commit; it's now like this:
```
if (nPos + bufsize < nSrcPos) {
```
This says that the distance from the new (requested) read position, `nPos`, to the source pointer, `nSrcPos` must be no greater than `bufsize`. If `nPos` is further back than that, it will be pointing to bytes from a _later_ write pass across the buffer and will get data that's too new. Maybe a small example will help. Suppose the file consists of `a b c d e f g ...` and the buffer is 5 bytes. After reading `g` (total of 7 bytes) from the file into the ring buffer, the buffer looks like this:
```
[ f g c d e ]
      ^
      nSrcPos = 7
```
The `nReadPos` (file read position) can be in the range 2 through 7 and we will get correct data. If `nReadPos` is _less_ than 2, let's say it's 1, then reading at that position should return `b` but will instead will return `g`, which is incorrect (`g` is too new). The smallest `nReadPos` can be is 2, which is 7 - 5 (`nSrcPos` - `bufsize`).

The existing `SetPos()` is wrong, `nRewind` has nothing to do with this. That variable is only needed during filling, and prevents `nSrcPos` from getting too far ahead of `nReadPos`, because if that happens, then `nReadPos` can't back up. So, in this little example,  if `nRewind` is 2, then `Fill()` will not let `nSrcPos` to get more than 5 (buffer size - rewind size) ahead of `nReadPos`. But, again, `nRewind` has no relevance to `SetPos()`.

Reply:
I also think that the first check in `SetPos()` is incorrect. This does not lead to problems in production thanks to the initialization values of `CBufferedFile` `vchBuf.size()=2*MAX_BLOCK_SERIALIZED_SIZE=8000000` and `nRewind=MAX_BLOCK_SERIALIZED_SIZE+8=4000008` in `LoadExternalBlockFile()`.

Accordingly, `nSrcPos = vchBuf.size() - nRewind = 3999992 > nRewind` after the first buffer fill and  the (erroneous) condition `nReadPos + nRewind < nSrcPos` in `SetPos()` is never `true`.

When I change  `LoadExternalBlockFile()` slightly by initializing CBufferedFile with `nRewind= MAX_BLOCK_SERIALIZED_SIZE-8` instead of `MAX_BLOCK_SERIALIZED_SIZE+8`, the functional test `feature_reindex` fails because now above condition can be `true`. Since all blocks in this test are much smaller than the maximum block size, this shouldn't happen.

Reply:
@mzumsande, it's broken even with the current production initialization values.  I made a temporary branch, https://github.com/LarryRuane/bitcoin/commit/bc9c987300782a7881edaabf7b7c5463a7688041, that simulates, after reading each block (during reindexing), what would happen if we really did need to rewind. (This temp branch doesn't include the changes from this PR.) `bitcoind --reindex` failed pretty quickly (at line 4405). 

The temp branch actually verifies two things, (1) the call to `SetPos()` returns `true` (the production code doesn't look at the return value), which means that the requested position was accommodated (as it should always be in the current context), and, (2), after rewinding, we get the expected data (by reading one byte).

I ran the same test with the code in the current PR's branch, and it doesn't fail.

I think the easiest way to understand why the current code is broken is to note that `Fill()` can push `nSrcPos` far ahead of `nReadPos`. For best efficiency,`Fill()` always reads as much as it can from the file in a single read, so if there's a lot of "runway" ahead of it (lots of space to the end of the buffer), it will read a lot. It may not go quite to the end of the buffer, if doing so would violate the rewind guarantee, but it can go much farther ahead of `nReadPos` than the rewind amount. Now look again at the current `SetPos()`:
```
if (nReadPos + nRewind < nSrcPos) {
    nReadPos = nSrcPos - nRewind;
    return false;
}
```
This condition could easily be true if `nSrcPos` has gotten far ahead of `nReadPos`. But that doesn't mean that `nReadPos` needs to be adjusted. Actually, this code could move `nReadPos` beyond the maximum position it's ever had, meaning that it's skipping data! You're at the farthest position you've ever reached, say, offset 40, and you request to rewind to position 35, and you end up at position 45! The condition is just wrong.

Reply:
You are right, I was more trying to understand why reindexing currently works in spite of the wrong condition. With different initialization values even no-rewind uses of `SetPos` like `SetPos(current_position)` could fail.